File: 2021/08/07/main.go
package main

func minCut(s string) int {
	n := len(s)
	l := make([]int, n+1)
	dp := make([][]bool, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = make([]bool, n)
		for j := 0; j <= n; j++ {
			dp[i][j] = false
		}
	}

	for j := 0; j < n; j++ {
		for i := j; i > -1; i-- {
			if s[i] == s[j] && (j-i < 2 || dp[i+1][j]) {
				dp[i][j] = true
				if l[j] > l[i]+1 {
					l[j] = l[i] + 1
				}
			}
		}
	}

	return l[0]
}

File: 2021/08/07/main.py
class Solution:
    def minCut(self, s: str) -> int:
        n = len(s)
        l = [i for i in range(-1, n)]
        dp = [[False]*n for i in range(n)]
        
        for j in range(n):
            for i in range(j, -1, -1):
                if (s[i] == s[j]) and (j-i < 2 or dp[i+1][j-1]):
                    dp[i][j] = True
                    cuts[j+1] = min(l[j+1], l[i]+1)
                    
        return l[-1]